File: deck/seed/code/native/browser/runtime/cryptography.ts
// Cryptography runtime for the Web platform. Digest names are normalised from the seed spelling (`sha256`) to the Web
// Crypto spelling (`SHA-256`). `digest` is async because `crypto.subtle` is. Reached only through the public
// cryptography API.
const cryptography = {
  digest: async (algorithm: string, data: string): Promise<string> => {
    const names: Record<string, string> = {
      sha256: 'SHA-256',
      sha384: 'SHA-384',
      sha512: 'SHA-512',
      sha1: 'SHA-1',
    }

    const name = names[algorithm] ?? algorithm.toUpperCase()
    const encoded = new TextEncoder().encode(data)
    const hash = await crypto.subtle.digest(name, encoded)

    return Array.from(new Uint8Array(hash))
      .map(byte => byte.toString(16).padStart(2, '0'))
      .join('')
  },
  randomBytes: (size: number): Uint8Array =>
    crypto.getRandomValues(new Uint8Array(size)),
}

File: deck/seed/code/native/kotlin/runtime/cryptography.kt
// Cryptography runtime. Digest names are normalised from the seed spelling (`sha256`) to the JCA spelling
// (`SHA-256`), which is the only reason this needs a shim rather than a direct call. Reached only through the public
// cryptography API.
object cryptography {
    private fun standardName(algorithm: String): String = when (algorithm) {
        "sha256" -> "SHA-256"
        "sha384" -> "SHA-384"
        "sha512" -> "SHA-512"
        "md5" -> "MD5"
        "sha1" -> "SHA-1"
        else -> algorithm.uppercase()
    }

    fun digest(algorithm: String, data: String): String {
        val digest = java.security.MessageDigest.getInstance(standardName(algorithm))
        val bytes = digest.digest(data.toByteArray(Charsets.UTF_8))

        return bytes.joinToString("") { "%02x".format(it) }
    }

    fun randomBytes(size: Long): ByteArray {
        val bytes = ByteArray(size.toInt())
        java.security.SecureRandom().nextBytes(bytes)

        return bytes
    }
}
